llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -4933,7 +4933,8 @@ InstCombinerImpl::pushFreezeToPreventPoisonFromPropagating(FreezeInst &OrigFI) {
   // poison.
   Value *MaybePoisonOperand = nullptr;
   for (Value *V : OrigOpInst->operands()) {
-    if (isa<MetadataAsValue>(V) || isGuaranteedNotToBeUndefOrPoison(V) ||
+    if (isa<MetadataAsValue>(V) ||
+        isGuaranteedNotToBeUndefOrPoison(V, &AC, &OrigFI, &DT) ||
```

Comment:
Given this is now passing in the assumption cache we should be able to remove freeze in more places if there is an assume instruction associated with one of the operands. It would be good to add a test for this I think.